keps/sig-windows/20190424-windows-cri-containerd.md

```diff
@@ -151,34 +155,284 @@ Operators could deploy additional RuntimeClasses with more granular control for
 
-
 #### Enable runtime resizing of container resources
 
 With virtual-based allocations and Hyper-V, it should be possible to increase the limit for a running pod. This won’t give it a guaranteed allocation, but will allow it to grow without terminating and scheduling a new pod. This could be a path to vertical pod autoscaling. This still needs more investigation and is mentioned as a future possibility.
 
-
 ### Implementation Details/Notes/Constraints
 
 The work needed will span multiple repos, SIG-Windows will be maintaining a [Windows CRI-Containerd Project Board] to track everything in one place.
 
-#### Proposal: Use Runtimeclass Scheduler to simplify deployments based on OS version requirements
-
-As of version 1.14, RuntimeClass is not considered by the Kubernetes scheduler. There’s no guarantee that a node can start a pod, and it could fail until it’s scheduled on an appropriate node. Additional node labels and nodeSelectors are required to avoid this problem. [RuntimeClass Scheduling](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/runtime-class-scheduling.md) proposes being able to add nodeSelectors automatically when using a RuntimeClass, simplifying the deployment.
-
-Windows forward compatibility will bring a new challenge as well because there are two ways a container could be run:
-- Constrained to the OS version it was designed for, using process-based isolation
-- Running on a newer OS version using Hyper-V.
-This second case could be enabled with a RuntimeClass. If a separate RuntimeClass was used based on OS version, this means the scheduler could find a node with matching class.
+#### Options for supporting multiple OS versions
+
+##### Option A - Use RuntimeClass per Windows OS version
+
+With process isolation, Windows containers must be constrained to only run on a matching Windows Server version node.
+
+[RuntimeClass Scheduling] has a few useful features that can make it easier to match a Windows container to a compatible Windows node. Here's an example of steps that could use RuntimeClass to avoid os version mismatches: 
+
+1. A RuntimeClass such as `WindowsServer1809` defined with [RuntimeClass Scheduling] `nodeSelector` for `kubernetes.io/os=windows` and `nodeSelector` for `beta.kubernetes.io/osversion=Windows1809` set.
+1. Nodes must be labeled with the `kubernetes.io/os=windows` (already there), as well as a new label for the OS version such as `beta.kubernetes.io/osversion=Windows1809`
+
+This would need to be repeated for each Windows version added to the cluster such as `windows-1809`, `windows-1903` and future releases. The cluster operator would need to ensure that nodes are added and scaled as needed for each Windows version independently.
+
+```yaml
+apiVersion: node.k8s.io/v1beta1
+kind: RuntimeClass
+metadata:
+  name: windows-1809-amd64
+handler: 'docker'  # only handler accepted when using dockershim
+scheduling:
+  nodeSelector:
+    kubernetes.io/os: 'windows'
```

Comment:
todo: open mini kep for os.version. that's needed regardless of this

Reply:
If kubelet needs to query the Windows OS version through something like syscall.GetVersion, we will also need to make sure there is a manifest generated: https://github.com/golang/go/issues/17835#issuecomment-371671658